_posts/please-welcome-model-binding-formatters-to-servicecomposer.md

```diff
@@ -0,0 +1,156 @@
+
+```csharp
+class RequestModel
+{
+    [FromRoute] public int id { get; set; }
+    [FromBody] public BodyModel Body { get; set; }
+}
+```
+
+The `ReqestModel` class is what glues all the concepts that we need. We have a public property, id, that matches the route value key and is decorated with the `[FromRoute]` attribute to explain to the binder where the value comes from, and public property for the body using binding attributes as well.
+
+> By the way, class names can be whatever you want.
+
+With the two above classes in place, we can now write the following handler:
+
+```csharp
+[HttpPost("sample/{id}")]
+public async Task Handle(HttpRequest request)
+{
+    var requestModel = await request.Bind<RequestModel>();
+    var body = requestModel.Body;
+    var aString = body.AString;
+    var id = requestModel.id;
+
+    //use values as needed
+}
+```
+
+The presented code, thanks to model binding, becomes much more maintainable and easy to understand. Not to mention that it is now compiler safe, being strongly typed.
+
+## Formatters
+
+The welcomed side effect of using the ASP.Net model binding infrastructure is that now ServiceComposer automatically supports input formatters. Previously it was up to the user to determine, in their handling code, the serialization format of the incoming payload, with model binding that becomes transparent and can be configured using standard ASP.Net configuration and HTTP headers to describe the request content type.
+
+Similarly, we decided to support output formatters. Output formatters were a natural choice to decouple ServiceComposer from the hardcoded Json serialization format and start honoring the "accept" HTTP header. For backward compatibility reasons, output formatters are not enabled by default; they need to be explicitly enabled at configuration time:
+
+```csharp
+public void ConfigureServices(IServiceCollection services)
+{
+    services.AddViewModelComposition(options =>
+    {
+        options.ResponseSerialization.UseOutputFormatters = true;
+    });
+    services.AddControllers();
+}
+```
+
+Both formatters and model binding requires the MVC infrastructure to be in place; this means that, as in the above snippet, one of the following needs to be configured:
+
+- AddControllers()
+- AddControllersAndViews()
+- AddMvc
+- AddRazorPages()
+
+## Conclusion
+
+With model binding and input and output formatters, ServiceComposer is moving from providing only a low-level API to a more robust and easy-to-use approach to maximize the developer experience and reduce errors. Model binding and formatters pave the way for a feature that has been in the back of my mind since the inception: a controller-like API that allows users to express their binding intentions through the handle method signature. The road is still long, but the journey just started.
```

Comment:
```suggestion
With model binding and input and output formatters, ServiceComposer is moving from providing only a low-level API to a more robust and easy-to-use approach to maximize the developer experience and reduce errors. Model binding and formatters pave the way for a feature that has been in the back of my mind since its inception: a controller-like API that allows users to express their binding intentions through the handle method signature. The road is still long, but the journey just started.
```